File: libgomp/testsuite/libgomp.c/taskloop-4.c
/* { dg-do run } */
/* { dg-options "-O2 -fopenmp" } */

int u[64], v;

__attribute__((noinline, noclone)) int
test (int a, int b, int c, int d, void (*fn) (int, int, int, int),
      int *num_tasks, int *min_iters, int *max_iters)
{
  int i, t = 0;
  __builtin_memset (u, 0, sizeof u);
  v = 0;
  fn (a, b, c, d);
  *min_iters = 0;
  *max_iters = 0;
  *num_tasks = v;
  if (v)
    {
      *min_iters = u[0];
      *max_iters = u[0];
      t = u[0];
      for (i = 1; i < v; i++)
	{
	  if (*min_iters > u[i])
	    *min_iters = u[i];
	  if (*max_iters < u[i])
	    *max_iters = u[i];
	  t += u[i];
	}
    }
  return t;
}

void
grainsize (int a, int b, int c, int d)
{
  int i, j = 0, k = 0;
  #pragma omp taskloop firstprivate (j, k) grainsize(d)
  for (i = a; i < b; i += c)
    {
      if (j == 0)
	{
	  #pragma omp atomic capture
	    k = v++;
	  if (k >= 64)
	    __builtin_abort ();
	}
      u[k] = ++j;
    }
}

void
num_tasks (int a, int b, int c, int d)
{
  int i, j = 0, k = 0;
  #pragma omp taskloop firstprivate (j, k) num_tasks(d)
  for (i = a; i < b; i += c)
    {
      if (j == 0)
	{
	  #pragma omp atomic capture
	    k = v++;
	  if (k >= 64)
	    __builtin_abort ();
	}
      u[k] = ++j;
    }
}

int
main ()
{
  #pragma omp parallel
    #pragma omp single
      {
	int min_iters, max_iters, ntasks;
	/* If grainsize is present, # of task loop iters is >= grainsize && < 2 * grainsize,
	   unless # of loop iterations is smaller than grainsize.  */
	if (test (0, 79, 1, 17, grainsize, &ntasks, &min_iters, &max_iters) != 79
	    || min_iters < 17 || max_iters >= 17 * 2)
	  __builtin_abort ();
	if (test (-49, 2541, 7, 28, grainsize, &ntasks, &min_iters, &max_iters) != 370
	    || min_iters < 28 || max_iters >= 28 * 2)
	  __builtin_abort ();
	if (test (7, 21, 2, 15, grainsize, &ntasks, &min_iters, &max_iters) != 7
	    || ntasks != 1 || min_iters != 7 || max_iters != 7)
	  __builtin_abort ();
	/* If num_tasks is present, # of task loop iters is min (# of loop iters, num_tasks).  */
	if (test (-51, 2500, 48, 9, num_tasks, &ntasks, &min_iters, &max_iters) != 54
	    || ntasks != 9)
	  __builtin_abort ();
	if (test (0, 25, 2, 17, num_tasks, &ntasks, &min_iters, &max_iters) != 13
	    || ntasks != 13)
	  __builtin_abort ();
      }
  return 0;
}

File: libgomp/testsuite/libgomp.fortran/taskloop4.f90
! { dg-do run }
! { dg-options "-O2" }

  integer, save :: u(64), v
  integer :: min_iters, max_iters, ntasks, cnt
  procedure(grainsize), pointer :: fn
  !$omp parallel
  !$omp single
    fn => grainsize
    ! If grainsize is present, # of task loop iters is
    ! >= grainsize && < 2 * grainsize,
    ! unless # of loop iterations is smaller than grainsize.
    call test (0, 79, 1, 17, fn, ntasks, min_iters, max_iters, cnt)
    if (cnt .ne. 79) call abort
    if (min_iters .lt. 17 .or. max_iters .ge. 17 * 2) call abort
    call test (-49, 2541, 7, 28, fn, ntasks, min_iters, max_iters, cnt)
    if (cnt .ne. 370) call abort
    if (min_iters .lt. 28 .or. max_iters .ge. 28 * 2) call abort
    call test (7, 21, 2, 15, fn, ntasks, min_iters, max_iters, cnt)
    if (cnt .ne. 7) call abort
    if (min_iters .ne. 7 .or. max_iters .ne. 7) call abort
    if (ntasks .ne. 1) call abort
    fn => num_tasks
    ! If num_tasks is present, # of task loop iters is
    ! min (# of loop iters, num_tasks).
    call test (-51, 2500, 48, 9, fn, ntasks, min_iters, max_iters, cnt)
    if (cnt .ne. 54 .or. ntasks .ne. 9) call abort
    call test (0, 25, 2, 17, fn, ntasks, min_iters, max_iters, cnt)
    if (cnt .ne. 13 .or. ntasks .ne. 13) call abort
  !$omp end single
  !$omp end parallel
contains
  subroutine grainsize (a, b, c, d)
    integer, intent (in) :: a, b, c, d
    integer :: i, j, k
    j = 0
    k = 0
    !$omp taskloop firstprivate (j, k) grainsize (d)
    do i = a, b - 1, c
      if (j .eq. 0) then
        !$omp atomic capture
          k = v
          v = v + 1
        !$omp end atomic
        if (k .ge. 64) call abort
      end if
      j = j + 1
      u(k + 1) = j
    end do
  end subroutine grainsize
  subroutine num_tasks (a, b, c, d)
    integer, intent (in) :: a, b, c, d
    integer :: i, j, k
    j = 0
    k = 0
    !$omp taskloop firstprivate (j, k) num_tasks (d)
    do i = a, b - 1, c
      if (j .eq. 0) then
        !$omp atomic capture
          k = v
          v = v + 1
        !$omp end atomic
        if (k .ge. 64) call abort
      end if
      j = j + 1
      u(k + 1) = j
    end do
  end subroutine num_tasks
  subroutine test (a, b, c, d, fn, num_tasks, min_iters, max_iters, cnt)
    integer, intent (in) :: a, b, c, d
    procedure(grainsize), pointer :: fn
    integer, intent (out) :: num_tasks, min_iters, max_iters, cnt
    integer :: i
    u(:) = 0
    v = 0
    cnt = 0
    call fn (a, b, c, d)
    min_iters = 0
    max_iters = 0
    num_tasks = v
    if (v .ne. 0) then
      min_iters = minval (u(1:v))
      max_iters = maxval (u(1:v))
      cnt = sum (u(1:v))
    end if
  end subroutine test
end
